File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.cc
#include <iostream>
#include <vector>
#include <algorithm>
#define endl '\n'
#define fastIO() ios::sync_with_stdio(0),cin.tie(0),cout.tie(0)
#define For(i,a,b) for(int i=a;i<b;i++)
using namespace std;

typedef long long ll;

struct Edge {
	int from;
	int to;
	ll weight;

	Edge(int from, int to, ll weight) :from(from), to(to), weight(weight) {}

	bool operator < (Edge e) {
		return weight < e.weight;
	}
};

ll V, E, A, B, C;

vector<Edge> edgeList;

int parents[10001];

void makeSet() {
	For(i, 1, V + 1) {
		parents[i] = i;
	}
}

int find(int a) {
	
	if (parents[a] == a) return a;

	return parents[a] = find(parents[a]);

}



bool uni(int a,int b) {
	int aRoot = find(a);
	int bRoot = find(b);

	if (aRoot == bRoot) return false;
	
	parents[bRoot] = aRoot;
	return true;

}


int main() {
	fastIO();
	cin >> V >> E;
	For(i, 0, E) {
		cin >> A >> B >> C;
		edgeList.push_back(Edge(A, B, C));
	}


	makeSet();
	sort(edgeList.begin(), edgeList.end());

	int cnt = 0;
	ll sm = 0;

	for (Edge e : edgeList) {
		if (!uni(e.from, e.to)) continue;
		sm += e.weight;
		cnt++;
		if (cnt == V-1) break;
	}
	cout << sm << endl;



	return 0;
}

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.io.*;
import java.util.*;
import java.awt.*;

class Main {
	static BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static int V,E,A,B;
	static long C,ans;
	static class Edge implements Comparable<Edge>{
		
		int from;
		int to;
		long weight;
		
		Edge(int from, int to,long weight){
			this.from = from;
			this.to = to;
			this.weight = weight;
		}
		
		public int compareTo(Edge o) {
			return Long.compare(weight, o.weight);
		}
	}
	static void init() {
		for(int i=1;i<=V;i++) {
			parents[i] =i;
		}
	}
	static int find(int a) {
		if(a == parents[a]) return a;
		else return parents[a] = find(parents[a]);
	}
	static boolean union(int a,int b) {
		int aRoot = find(a);
		int bRoot = find(b);
		if(aRoot == bRoot) return false;
		else {
			parents[bRoot] = aRoot;
			return true;
		}
	}
	static int[] parents;
	
	static ArrayList<Edge> edgeList;
	public static void main(String[] args) throws IOException {
		st = new StringTokenizer(in.readLine());
		V = Integer.parseInt(st.nextToken());
		E = Integer.parseInt(st.nextToken());
		
		parents = new int[V+1];
		edgeList = new ArrayList<>();
		for(int i=0;i<E;i++) {
			st = new StringTokenizer(in.readLine());
			A = Integer.parseInt(st.nextToken());
			B = Integer.parseInt(st.nextToken());
			C = Long.parseLong(st.nextToken());
			edgeList.add(new Edge(A,B,C));
		}
		init();
		
		Collections.sort(edgeList);
		int cnt =0;
		
		for(Edge e : edgeList ) {
			if(!union(e.from,e.to)) continue;
			ans += e.weight;
			
			if(++cnt == V-1 ) break;
		}
		System.out.println(ans);
		
		
	}
}